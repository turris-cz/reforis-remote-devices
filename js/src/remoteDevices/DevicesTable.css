.devices-table td {
    vertical-align: middle;
}

@media (max-width: 1400px) {
    .devices-table-delete-icon {
        display: none;
    }
}
