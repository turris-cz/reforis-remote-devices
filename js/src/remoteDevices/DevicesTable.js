/*
 * Copyright (C) 2019-2023 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";

import { Button, useWSForisModule } from "foris";
import PropTypes from "prop-types";

import EditableName from "./editableName/EditableName";
import ToggleDevice from "./ToggleDevice";

import "./DevicesTable.css";

const deviceShape = PropTypes.shape({
    controller_id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    options: PropTypes.shape({
        custom_name: PropTypes.string,
    }),
});

DevicesTable.propTypes = {
    ws: PropTypes.object.isRequired,
    devices: PropTypes.arrayOf(deviceShape).isRequired,
    deleteDevice: PropTypes.func.isRequired,
    patchDevice: PropTypes.func.isRequired,
};

function DevicesTable({ ws, devices, deleteDevice, patchDevice }) {
    if (!devices || devices.length === 0) {
        return (
            <p className="text-muted text-center">
                {_("No devices added yet.")}
            </p>
        );
    }

    return (
        // Bottom padding is required to prevent unnecessary vertical scroll when editor is active
        <div className="table-responsive pb-2">
            <table className="table table-hover devices-table">
                <thead className="thead-light">
                    <tr>
                        <th scope="col">{_("ID")}</th>
                        <th scope="col">{_("Name")}</th>
                        <th scope="col" className="text-center">
                            {_("Status")}
                        </th>
                        <th scope="col" className="text-center">
                            {_("Managed")}
                        </th>
                        <th
                            scope="col"
                            className="text-right"
                            aria-label={_("Delete")}
                        />
                    </tr>
                </thead>
                <tbody>
                    {devices.map((device) => (
                        <DeviceRow
                            key={device.controller_id}
                            ws={ws}
                            device={device}
                            deleteDevice={() => {
                                deleteDevice({ suffix: device.controller_id });
                            }}
                            patchDevice={(data) => {
                                patchDevice({
                                    data,
                                    suffix: device.controller_id,
                                });
                            }}
                        />
                    ))}
                </tbody>
            </table>
        </div>
    );
}

export default DevicesTable;

DeviceRow.propTypes = {
    ws: PropTypes.object.isRequired,
    device: deviceShape,
    deleteDevice: PropTypes.func.isRequired,
    patchDevice: PropTypes.func.isRequired,
};

function DeviceRow({ ws, device, deleteDevice, patchDevice }) {
    const [status, setStatus] = useState();

    const [advertizeNotification] = useWSForisModule(ws, "remote", "advertize");
    useEffect(() => {
        if (
            advertizeNotification &&
            advertizeNotification.id === device.controller_id
        ) {
            setStatus(advertizeNotification.state);
        }
    }, [advertizeNotification, device.controller_id, status]);

    return (
        <tr>
            <td>{device.controller_id}</td>
            <td className="editable-name">
                <EditableName
                    name={device.options.custom_name}
                    patchDevice={patchDevice}
                />
            </td>
            <td className="text-center">
                {/* "key" is necessary to update the icon, see https://stackoverflow.com/q/47722813/6324591 */}
                <StatusIcon
                    status={status}
                    key={`${device.controller_id}-${status}`}
                />
            </td>
            <td className="text-center">
                <ToggleDevice
                    controllerID={device.controller_id}
                    enabled={device.enabled}
                    patchDevice={patchDevice}
                />
            </td>
            <td className="text-right">
                <Button className="btn-sm btn-danger" onClick={deleteDevice}>
                    <i className="fa fa-trash-alt mr-2 devices-table-delete-icon" />
                    {_("Delete")}
                </Button>
            </td>
        </tr>
    );
}

StatusIcon.propTypes = {
    status: PropTypes.string,
};

function StatusIcon({ status }) {
    let className = "fa-question-circle text-warning";
    let statusDescription = _("Unknown status");
    if (status === "started") {
        className = "fa-play-circle text-primary";
        statusDescription = _("Started");
    } else if (status === "running") {
        className = "fa-check-circle text-success";
        statusDescription = _("Running");
    } else if (status === "exitted") {
        className = "fa-times-circle text-danger";
        statusDescription = _("Exited");
    }

    /*
     * Wrapper tag is required to properly remove icon because "i" element
     * is actually replaced by "svg" element.
     */
    return (
        <span>
            <i className={`fa fa-lg ${className}`} title={statusDescription} />
        </span>
    );
}
