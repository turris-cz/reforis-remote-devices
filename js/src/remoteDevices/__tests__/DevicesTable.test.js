/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, fireEvent } from "foris/testUtils/customTestRender";

import DevicesTable from "../DevicesTable";
import devicesFixture from "./__fixtures__/devices";

describe("<DevicesTable />", () => {
    const onDelete = jest.fn();

    function renderTable(devices, isDeleting = false) {
        return render(
            <DevicesTable devices={devices} isDeleting={isDeleting} onDelete={onDelete} />,
        );
    }

    it("should handle empty devices list", () => {
        const { container } = renderTable([]);
        expect(container).toMatchSnapshot();
    });

    it("should render devices table", () => {
        const { container } = renderTable(devicesFixture);
        expect(container).toMatchSnapshot();
    });

    it("should execute onDelete callback when button is pressed", () => {
        const { getAllByText } = renderTable(devicesFixture);
        const deleteButtons = getAllByText("Delete");
        fireEvent.click(deleteButtons[0]);
        expect(onDelete).toBeCalledWith({ suffix: devicesFixture[0].controller_id });
    });
});
