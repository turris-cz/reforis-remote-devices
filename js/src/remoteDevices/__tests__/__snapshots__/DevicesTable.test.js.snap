// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DevicesTable /> should handle empty devices list 1`] = `
<div>
  <p
    class="text-muted text-center"
  >
    No devices added yet.
  </p>
</div>
`;

exports[`<DevicesTable /> should render devices table 1`] = `
<div>
  <div
    class="table-responsive pb-2"
  >
    <table
      class="table table-hover devices-table"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            ID
          </th>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            class="text-center"
            scope="col"
          >
            Status
          </th>
          <th
            class="text-center"
            scope="col"
          >
            Managed
          </th>
          <th
            aria-label="Delete"
            class="text-right"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            A1
          </td>
          <td
            class="editable-name"
          >
            <form
              class="editable-name-form"
            >
              <div
                class="editable-name-input-wrapper"
              >
                <input
                  class="editable-name-input-disabled"
                  maxlength="30"
                  placeholder="Custom name"
                  readonly=""
                  type="text"
                  value="A"
                />
              </div>
              <button
                class="btn btn-link"
                title="Edit name"
                type="button"
              >
                <i
                  class="fa fa-edit text-muted"
                />
              </button>
            </form>
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-question-circle text-warning"
                title="Unknown status"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                checked=""
                class="custom-control-input"
                id="toggle-A1"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="toggle-A1"
              >
                Yes
              </label>
            </div>
          </td>
          <td>
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2 devices-table-delete-icon"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            A2
          </td>
          <td
            class="editable-name"
          >
            <form
              class="editable-name-form"
            >
              <div
                class="editable-name-input-wrapper"
              >
                <input
                  class="editable-name-input-disabled"
                  maxlength="30"
                  placeholder="Custom name"
                  readonly=""
                  type="text"
                  value="A"
                />
              </div>
              <button
                class="btn btn-link"
                title="Edit name"
                type="button"
              >
                <i
                  class="fa fa-edit text-muted"
                />
              </button>
            </form>
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-question-circle text-warning"
                title="Unknown status"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                class="custom-control-input"
                id="toggle-A2"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="toggle-A2"
              >
                No
              </label>
            </div>
          </td>
          <td>
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2 devices-table-delete-icon"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            B1
          </td>
          <td
            class="editable-name"
          >
            <form
              class="editable-name-form"
            >
              <div
                class="editable-name-input-wrapper"
              >
                <input
                  class="editable-name-input-disabled"
                  maxlength="30"
                  placeholder="Custom name"
                  readonly=""
                  type="text"
                  value="B"
                />
              </div>
              <button
                class="btn btn-link"
                title="Edit name"
                type="button"
              >
                <i
                  class="fa fa-edit text-muted"
                />
              </button>
            </form>
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-question-circle text-warning"
                title="Unknown status"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                checked=""
                class="custom-control-input"
                id="toggle-B1"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="toggle-B1"
              >
                Yes
              </label>
            </div>
          </td>
          <td>
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2 devices-table-delete-icon"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            B2
          </td>
          <td
            class="editable-name"
          >
            <form
              class="editable-name-form"
            >
              <div
                class="editable-name-input-wrapper"
              >
                <input
                  class="editable-name-input-disabled"
                  maxlength="30"
                  placeholder="Custom name"
                  readonly=""
                  type="text"
                  value="B"
                />
              </div>
              <button
                class="btn btn-link"
                title="Edit name"
                type="button"
              >
                <i
                  class="fa fa-edit text-muted"
                />
              </button>
            </form>
          </td>
          <td
            class="text-center"
          >
            <span>
              <i
                class="fa fa-lg fa-question-circle text-warning"
                title="Unknown status"
              />
            </span>
          </td>
          <td
            class="text-center"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                class="custom-control-input"
                id="toggle-B2"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="toggle-B2"
              >
                No
              </label>
            </div>
          </td>
          <td>
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2 devices-table-delete-icon"
              />
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
